### ISSUE TITLE
bad rendering and network.body.view.translation is {x:NaN, y:NaN}

### ISSUE THREAD

**Issue by nicolaslupinski** [visjs/vis-network#12]
Hi

There is a hard-to-reproduce bug with vis network 4.21.0 : sometimes, the networks is not rendered properly (only edges), and it is related to network.body.view.translation being {x:NaN, y:NaN}.

I'm getting the issue on all kind of browser and on different computers, especially the slower one. Maybe the browser is trying to render before the network has been initialized ?

I've found I can reset the view translation to x:0,y:0, then redraw().

Is this bug solved in the 5.0 ?

Regards

**Comment by mojoaxel** [comment]
> Is this bug solved in the 5.0 ? 

You tell us :wink: 

Please try it with the 5.0 release. If there the problem still exists you also can try [visjs-network](https://github.com/visjs-community/visjs-network). If the problem is fixed there we might adopt those changes.

**Comment by mojoaxel** [comment]
@nicolaslupinski Can you please confirm this problem still exists in v5.1.0 !?

**Comment by Thomaash** [comment]
I actually encountered this too, very rarely though, with the original vis package. But it was happening only with Vue.js dev tools active so I blamed it on HMR or something. However all these tools slow everything down a lot, maybe it would also happen on a slow PC (if it is indeed the same issue).

**Comment by Thomaash** [comment]
@nicolaslupinski Does this ring a bell?
![Screenshot from 2019-07-29 13-21-07](https://user-images.githubusercontent.com/12966757/62044654-c4540f00-b203-11e9-8532-1ef6581e3da2.png)

(vis 4.21.0)

**Comment by Thomaash** [comment]
I'll try to trigger this with the new vis-network.

**Comment by Thomaash** [comment]
![Screenshot from 2019-07-29 13-33-28](https://user-images.githubusercontent.com/12966757/62045423-8e178f00-b205-11e9-871a-8ecd9fcc9cdf.png)


(vis-network 5.1.0)

**Comment by Thomaash** [comment]
Steps to reproduce:
- `net.moveTo({ position: { x, y }, scale })`
- `net.fit({ animation: true })`

**Comment by Thomaash** [comment]
> Steps to reproduce:
> 
> * `net.moveTo({ position: { x, y }, scale })`
> * `net.fit({ animation: true })`

Nope, it's more complex than that. This works just fine in MWE.

**Comment by Thomaash** [comment]
MWE: https://jsfiddle.net/6kg4xr7d/2/

**Comment by Thomaash** [comment]
The problem is that string scale (notice line 76 in MWE) doesn't work but also doesn't show any error.

**Comment by Thomaash** [comment]
Well it sorta works, but causes problems.

**Comment by nicolaslupinski** [comment]
> > Steps to reproduce:
> > 
> > * `net.moveTo({ position: { x, y }, scale })`
> > * `net.fit({ animation: true })`
> 

Thanks Thomaash !

I can reproduce the issue reliably by setting the scale to 0 (which make the whole network disappear) then network.fit().

network.moveTo({position:{x:1, y:1}, scale:0}); network.fit();

The hacked fix, to reset the network.body.view.translation before redraw and fit, also works reliably :

function redrawAndFit(network){
    if(isNaN(network.body.view.translation.x) || isNaN(network.body.view.translation.y)){
      network.body.view.translation = {x:0,y:0};
    }
    network.redraw();
    network.fit();
}


I'll have a look at the source, now.

**Comment by Thomaash** [comment]
So far two reasons have been reliably identified:
- string scale in moveTo: https://jsfiddle.net/6kg4xr7d/2/
- 0 scale in moveTo: https://jsfiddle.net/pq8heo9b/

**Comment by Thomaash** [comment]
@nicolaslupinski Will your problems disappear if you convert your scale to a number and ensure it is not zero?

**Comment by Thomaash** [comment]
@mojoaxel This seems like input validation issue to me. We should just throw if the input doesn't pass `scale > 0` to not cause this hard to debug confusion. Or is there any point in having zero, negative or string scale?

**Comment by mojoaxel** [comment]
> We should just throw if the input doesn't pass scale > 0 to not cause this hard to debug confusion. Or is there any point in having zero, negative or string scale?

I looked at it and I think you are right! The docs state that `The scale is a number greater than 0`. If you want you can add a Number cast bevor you check is `scale > 0` this way strings would also work. 

@Thomaash Could you provide a fix or should I do it?

**Comment by Thomaash** [comment]
I'll fix it.

**Comment by nicolaslupinski** [comment]
> @nicolaslupinski Will your problems disappear if you convert your scale to a number and ensure it is grater than zero?

I doubt.

My original problem is that the rendering is wrong sometimes.

Setting the scale to 0 was a reliable way to get the same bug (and test the fix). But it might not be the cause of the original problem (view.translation = {x:NaN, y:NaN}).

I'd rather have something that prevent view.translation to be set to NaN. I'll have to read the code to figure out where this can happen.

**Comment by nicolaslupinski** [comment]
I manage to reproduce by realoading again and again untill it triggers. There is more than one bug actually, sometimes I get : 

network.body.view.translation
{x: -Infinity, y: -Infinity}
network.body.view.scale
Infinity

Then I redraw() and and fit() :

network.body.view.translation
{x: NaN, y: NaN}
network.body.view.scale
0.8912253390328385

The rendering is messy : in the first case, it draws edges only, in the second it draw edges and their labels, but still no nodes.

**Comment by Thomaash** [comment]
@nicolaslupinski Hi, it's perfectly possible that I misidentified your issue and fixed an unrelated problem. Could you please adapt this MWE https://codesandbox.io/embed/vis-networkissues12-mwe-1-exct0 to show your issue so that we all can see it?

**Comment by nicolaslupinski** [comment]
Hi Thomaash

Sorry I don't provide feedback quickly.

I've edited the code to show my issue, but I feel i'm cheating, by reproducing the bug and not its original cause.

Regards

**Comment by Thomaash** [comment]
Hi @nicolaslupinski,

it seems that you forgot to share the link to your MWE. Could you provide it please?

Thanks.

**Comment by nicolaslupinski** [comment]
Indeed...

I boils down to writing  :

network.moveTo({
    position: {
      x: -Infinity,
      y: -Infinity
    },
    scale: 2
  });

**Comment by Thomaash** [comment]
@nicolaslupinski Your're right, this is still an issue.

**Comment by nicolaslupinski** [comment]
Hi

This bug seems more related to rendering in a complex web page.

I've tried to add 

`setTimeout( function(){ redrawAndFit(); redrawAndFit(); }, 1000);`

to my page, with

```
function redrawAndFit(){
   if(isNaN(network.body.view.translation.x) || isNaN(network.body.view.translation.y)){
      network.body.view.translation = {x:0,y:0}; //hack intended to fix bug
   }  
   network.redraw();
   network.fit();
}
```

But it is still unable to prevent the rendering bug from happening.

I've tried putting an alert() after rendering event, in the bugged case I'm getting the alert before the canva is even put on screen.

The network related code might be executed before all the proper initialisation have been performed, hence it might get undefined or 0 value for canva's height or width.
